bowtie: new versions and %gcc@8.0.0: fix

hartzell opened the pull request:
Bowtie development shifted from Sourceforge to GitHub.  This updates the Spack package to follow.

This commit adds several versions from GitHub, using the archive tarballs.  Note that the tag and URL for 1.2.2 tarball is actually 'v1.2.2_p1'...  Sigh.

Versions 1.2.0 and newer come from GitHub (I'm unsure if the sourceforge 1.2 and github 1.2.0 are identical).

It leaves the old 1.2 release download info in place.

The PR includes a fix from https://github.com/BenLangmead/bowtie/issues/87 that enables %gcc@8.0.0: 

Sadly, v1.2.2 has trouble with "newer" gcc's, so the PR it only adds v1.2.2 for %gcc@6.0.0:.  I'm not sure what gcc's work and/or don't.  Others should feel free to tighten this.  I know that:

- 1.2 -> 1.2.2 work with %gcc@5.5.0;
- 1.2 -> 1.2.1.1 work with %gcc@8.2.0; and
- 1.2.2 fails with %gcc@8.2.0

adamjstewart commented on var/spack/repos/builtin/packages/bowtie/package.py:
```diff
@@ -11,14 +11,31 @@ class Bowtie(MakefilePackage):
+    # - 1.2 -> 1.2.2 work with %gcc@5.5.0;
+    # - 1.2 -> 1.2.1.1 work with %gcc@8.2.0; and
+    # - 1.2.2 fails with %gcc@8.2.0
+    conflicts('%gcc@6.0.0:', when='@1.2.2')
```
I'd prefer to limit the conflict to versions that are _known_ to conflict. Obviously no one tries every package and every version with every compiler. I only add conflicts to things when I know they cause problems. If someone else finds a problem, they can extend the conflict.

hartzell replied:
So you'd suggest:

```python
conflicts('%gcc@8.2.0', when='@1.2.2')
```

adamjstewart replied:
I would also be fine with:
```python
conflicts('%gcc@8:', when='@1.2.2')
```

hartzell replied:
Sold!

hartzell commented on var/spack/repos/builtin/packages/bowtie/package.py:
```diff
@@ -11,14 +11,31 @@ class Bowtie(MakefilePackage):
+
+    # Release 1.2.2 fails to build with gcc@8.2.0
+    # Note that release 1.2.2 is tagged as v1.2.2_p1, so the URL is "odd".
+    version('1.2.2', sha256='e1b02b2e77a0d44a3dd411209fa1f44f0c4ee304ef5cc83f098275085740d5a1', url="https://github.com/BenLangmead/bowtie/archive/v1.2.2_p1.tar.gz")
```
Yes, which is why I wish they'd just released 1.2.3.  WTF....  Since when did integers become so precious???

The problem is that when normal biologists go and grab "v1.2.2", I'm fairly certain they're going to grab *this* one:

<img width="359" alt="screen shot 2018-12-31 at 10 06 46 pm" src="https://user-images.githubusercontent.com/312978/50570670-6b44fb00-0d48-11e9-9413-bb14358a4137.png">

So that's what I argue we should give them with our `1.2.2`

hartzell replied:
One more observation about this, if they download the prebuilt binaries for v1.2.2 from the releases page, they're getting things associated with the v1.2.2_p1 tag, not v1.2.2.

hartzell replied:
@adamjstewart -- are you ok with me leaving it the way it is?

adamjstewart replied:
Hmm, I see your point, but I also see my point. Let's wait for someone else to act as a tie breaker 😃

hartzell replied:
Suggestions on whom we might tag/pester?

adamjstewart replied:
@davydden and @alalazo usually have opinions on trivial things like this. Hopefully they don't both respond and both disagree 😛

alalazo replied:
> Since when did integers become so precious???
 
It could be worse. They could have patched 1.2.3 and make it 1.2.&#960; (there should be a penalty box for tagging versions like that :speak_no_evil: ) That said I see points in both approaches - maintaining the tag and removing it. I'd be slightly in favor of what @hartzell does here. Calling the version 1.2.2 is a lie, but one that makes your life easier (as no other version in the package has a tag).

davydden replied:
maybe to avoid the confusion within Spack, add the older `1.2.2` (the link @adamjstewart  found) and name this one `1.2.2_p1`?  I hope the presence of both will make the whole story a bit more clear for Spack users. I would still leave an incode comment about the whole issue.

davydden replied:
> Hopefully they don't both respond and both disagree

here you go, we do ;-)

adamjstewart replied:
@tgamblin can be our tie-breaker then. Or is there anyone else who actually uses `bowtie` and cares? It looks like @weijianwen wrote the package.

hartzell replied:
Ah, so @adamjstewart, you'd be good with spack's having a `1.2.2_p1` that points to the `v1.2.2_p1` tarball, and an explanatory note in the package/docs?

adamjstewart replied:
Yeah, I'm fine with that. I could also be convinced to keep things the way they currently are if someone breaks the tie. Sorry if I wasn't clear about what I was suggesting. I definitely don't think it's worth adding the original broken 1.2.2 tarball to the package.

hartzell replied:
Well, I thought that I was going to "give in", on the theory that bowtie has an ugly wart and all that we're doing it discussing how Spack's going to show that wart to the user.

Thought I'd test it though and wired the package up to use `v.1.2.2_p1` and just not mention `v1.2.2`.

The problem/feature with this approach is that a determined person *can* install what Spack gives them when they ask for **v.1.2.2** but it won't be the same as the prebuilt binaries or the uploaded release artifact/tarball that are available on the GitHub release page.  E.g.:

```console
[ghartzell@bifx1n03 spack]$ spack install bowtie@1.2.2
==> Installing bowtie
==> Searching for binary cache of bowtie
==> Finding buildcaches in /home/ghartzell/spacktree/mirror/build_cache
==> No binary for bowtie found: installing from source
==> Warning: There is no checksum on file to fetch bowtie@1.2.2 safely.
==>   Fetch anyway? [y/N] ^C
==> Error: Keyboard interrupt.
```

I suppose that I could add a `v1.2.2` like so and just trust that no one will ever "fix" the package to include the real tarball's signature:

```python
version('1.2.2', md5='deadbeefdeadbeefdeadbeefdeadbeef')
```

I'm not a fan.  I wouldn't care so much if the difference was a copyright, or a README, or ....  Sadly it's an off-by one in some code that loops over the number of threads...

I'm back to standing in my corner of the room, with my arms crossed and glaring balefully, hoping that someone breaks the tie in "my" favor.

😤

hartzell replied:
I guess a third possibility is to do this:

```python
# Note that the GitHub release 1.2.2 is actually git tagged as v1.2.2_p1.
# We point both of the Spack versions at the same tarball so they build the binaries
# that are on the release page as v1.2.2
version('1.2.2_p1', sha256='e1b02b2e77a0d44a3dd411209fa1f44f0c4ee304ef5cc83f098275085740d5a1')
version('1.2.2', sha256='e1b02b2e77a0d44a3dd411209fa1f44f0c4ee304ef5cc83f098275085740d5a1', url="https://github.com/BenLangmead/bowtie/archive/v1.2.2_p1.tar.gz")
```

adamjstewart replied:
You make a good point about Spack downloading `1.2.2` even without a checksum. I guess a counterpoint would be that `spack versions` and `spack checksum` will report a new version that doesn't exist in the package, even though it does. How many people actually say `spack install bowtie@1.2.2` instead of `spack install bowtie`?

hartzell replied:
> How many people actually say spack install bowtie@1.2.2 instead of spack install bowtie?

The various sites that I've influenced (3 or so) have a `packages.yaml` that pins the versions of the things that they're installing.  I can easily see (which is why I'm being such a pill about finding the least distasteful way to handle what the bowtie folks did) someone telling the computer-ish folks: "We need to use `v1.2.2`."  and having that be directly dropped into a `packages.yaml`.  Not sure what other folks are using.  I'm not sure what would actually encourage someone to go dig into the Releases and tags and *etc*. 

> [...] report a new version that doesn't exist in the package, even though it does. 

True.  Perhaps that's a reason for a big bunch of `deadbeef` and a comment in the package.  Spack would never merge a PR and if someone really, really, really needed the broken release they could edit their `.../bowtie/package.py`.

On a related note:

The binaries that one can download from the v1.2.2 Release section report themselves as *version 1.2.2*, even though they're (presumably) built from the `v1.2.2_p1` tag.

```console
(alice)[19:34:17]Desktop>>./bowtie-1.2.2-macos-x86_64/bowtie --version
/Users/hartzell/Desktop/bowtie-1.2.2-macos-x86_64/bowtie-align-s version 1.2.2
64-bit
Built on jupiter.local
Tue Dec 12 13:35:25 EST 2017
Compiler: InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
Options: -O3 -m64  -mmacosx-version-min=10.9 -DWITH_TBB -DPOPCNT_CAPABILITY
Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}
(alice)[19:34:26]Desktop>>
```

hartzell replied:
This effectively disables the 1.2.2 version

```python
# The bowtie project git tagged and GitHub released a v1.2.2,
# discovered/fixed a bug, git tagged a v1.2.2_p1 and moved the
# 1.2.2 release to use it.
#
# As a result the binaries, the uploaded tarball artifact and the
# GitHub generated archive files that are named 1.2.2 are actually
# built from the v1.2.2_p1 tag.  No one uses the v1.2.2 tag for
# anything.
#
# For our Spack versions, we're forcing people to be aware of this
# and ask for 1.2.2_p1 and we've made it tough to actually build
# from the 1.2.2 tag.
version('1.2.2_p1', sha256='e1b02b2e77a0d44a3dd411209fa1f44f0c4ee304ef5cc83f098275085740d5a1')
version('1.2.2', sha256='DO_NOT_USE_THIS_VERSION')
```

The "UX" looks like this:

```console
[ghartzell@bifx1n03 spack]$ spack install bowtie@1.2.2
==> Installing bowtie
==> Searching for binary cache of bowtie
==> Finding buildcaches in /home/ghartzell/spacktree/mirror/build_cache
==> No binary for bowtie found: installing from source
==> Fetching file:///home/ghartzell/spacktree/mirror/bowtie/bowtie-1.2.2.tar.gz
curl: (37) Couldn't open file /home/ghartzell/spacktree/mirror/bowtie/bowtie-1.2.2.tar.gz
==> Fetching from file:///home/ghartzell/spacktree/mirror/bowtie/bowtie-1.2.2.tar.gz failed.
==> Fetching https://github.com/BenLangmead/bowtie/archive/v1.2.2.tar.gz
            -=O=-                                  #            #             #            #
==> Error: ValueError: Spack knows no hash algorithm for this digest: DO_NOT_USE_THIS_VERSION

/home/ghartzell/spack/lib/spack/spack/package.py:965, in do_fetch:
        962        self.stage.fetch(mirror_only)
        963        self._fetch_time = time.time() - start_time
        964
  >>    965        if checksum and self.version in self.versions:
        966            self.stage.check()
        967
        968        self.stage.cache_local()

```

alalazo replied:
> I guess a third possibility is to do this: [...]

Would it be a sensible possibility to use `url_for_version` to print an info message that says the same thing as the comment if one installs `1.2.2`?

hartzell replied:
Seems like it's doable.  Would such a use of `url_for_version` be acceptable?

adamjstewart replied:
Hmm, seems hacky.

odoublewen replied:
@adamjstewart if you don't mind my comment from the peanut gallery:
> Or is there anyone else who actually uses bowtie and cares?

I actually use bowtie... After reading through this thread, my druthers would be @hartzell's ["third possibility"](https://github.com/spack/spack/pull/10232#discussion_r244902714).   Just point both `1.2.2` and `1.2.2_p1` to the same tarball.  Nobody will *really* want to install the broken 1.2.2 release.  I don't think it is the place of Spack to provide "rope to hang from".

Seems like any of these fixes will be a bit hacky under the hood, but from a user perspective, the above seems the least likely to create confusion.

hartzell replied:
Ok, both v1.2.2 adn v1.2.2_p1 point at the v1.2.2_p1 tarball.

Have we achieved consensus 🎊?

hartzell replied:
@adamjstewart -- you started this conversation and I think it's ended up at a place you're comfortable with.

Can we resolve it?

hartzell commented:
I'd like to get this resolved and put to bed.  @adamjstewart raised the last objection, which I think I've addressed.  He mentioned having @tgamblin weigh in, is that still necessary?

Quick summary: bowtie tagged and released 1.2.2, patched it, added a 1.2.2_p1 tag, and pointed the *old* release name and artifacts (1.2.2) at the new tag (1.2.2_p1).

We wrestled with how Spack should handle this and ended up pointing Spack versions 1.2.2 and 1.2.2_p1 at the 1.2.2_p1 tarball.

@odoublewen, who is part of a community I support, weighed in supporting this choice.

hartzell commented:
Thanks!  :smile:
